keras_nlp/models/gpt_neox/gpt_neox_attention.py

```diff
@@ -0,0 +1,177 @@
+        self._kernel_initializer = keras.initializers.get(kernel_initializer)
+        self._bias_initializer = keras.initializers.get(bias_initializer)
+
+        self._qkv_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_heads, 3 * self.attn_head_size),
+            bias_axes="de",
+            **self._get_common_kwargs_for_sublayer(use_bias=True),
+            name="query",
+        )
+
+        self._attn_dropout_layer = keras.layers.Dropout(
+            self.dropout, name="attention_dropout"
+        )
+
+        self._softmax = keras.layers.Softmax(axis=-1, name="attention_softmax")
+
+        # Output.
+        self._output_dense = keras.layers.EinsumDense(
+            equation="abc,cd->abd",
+            output_shape=(None, self.hidden_dim),
+            bias_axes="d",
+            **self._get_common_kwargs_for_sublayer(use_bias=True),
+            name="attention_output",
+        )
+
+    def _get_common_kwargs_for_sublayer(self, use_bias=True):
+        common_kwargs = {}
+
+        kernel_initializer = clone_initializer(self._kernel_initializer)
+        bias_initializer = clone_initializer(self._bias_initializer)
+
+        common_kwargs["kernel_initializer"] = kernel_initializer
+        if use_bias:
+            common_kwargs["bias_initializer"] = bias_initializer
+
+        return common_kwargs
+
+    def _masked_softmax(self, attention_scores, attention_mask=None):
+
+        if attention_mask is not None:
+            mask_expansion_axis = -3
+            for _ in range(
+                attention_scores.shape.rank - attention_mask.shape.rank
+            ):
+                attention_mask = tf.expand_dims(
+                    attention_mask, axis=mask_expansion_axis
+                )
+        return self._softmax(attention_scores, attention_mask)
+
+    def _compute_attention(
+        self, query, key, value, attention_mask=None, training=None
+    ):
+
+        attention_scores = tf.einsum("aecd,abcd->acbe", key, query)
+        attention_scores /= self.norm_factor
+
+        attention_scores = self._masked_softmax(
+            attention_scores, attention_mask
+        )
+        attention_scores = self._attn_dropout_layer(
+            attention_scores, training=training
+        )
+        attention_output = tf.einsum("acbe,aecd->abcd", attention_scores, value)
+
+        return attention_output, attention_scores
+
+    def call(
+        self,
+        hidden_states,
+        attention_mask,
+        layer_past=None,
+        return_attention_scores=False,
+        training=None,
+    ):
+
+        query_key_value = self._qkv_dense(hidden_states)
+
+        query = query_key_value[..., : self.attn_head_size]
+        key = query_key_value[
+            ..., self.attn_head_size : 2 * self.attn_head_size
+        ]
+        value = query_key_value[..., 2 * self.attn_head_size :]
+
+        query_rot, query_pass = (
```

Comment:
I wonder if we would be better off moving this slice and concat logic into the `RotaryEmbedding` call. Then our usage here could look a little more like...

```
query = self.rotary_embedding(query)
key = self.rotary_embedding(key)
```

And the rotary embedding layer could also hold the percentage argument, which would conceptually be quite clean. Looks like falcon is doing this roughly -> https://huggingface.co/tiiuae/falcon-40b/blob/main/modelling_RW.py

Reply:
Thanks for this wonderful suggestion !